### PR TITLE
Feature/932 China RP1 implementation

### PR DESCRIPTION
**Pull request by MaggieSalak**
# PR for issue #931 

## What is being addressed

- China frequency plan implementation for RP1 specification 
https://lora-alliance.org/wp-content/uploads/2020/11/lorawan_regional_parameters_v1.0.3reva_0.pdf

## How is this addressed

- New Region implemented (`RegionCN470RP1`)
- The existing China implementation renamed to `RegionCN470RP2`
- New `LoRaRegionType` added and the existing one was renamed
- Unit tests added to cover the new region

### CODE REVIEW

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/Regions/Region.cs [hunk 1]
```diff
@@ -267,6 +267,46 @@ public string GetDownstreamRX2DataRate(string devEUI, string nwkSrvRx2Dr, ushort
...
         }
 
+        /// <summary>
+        /// Get downstream RX2 data rate.
+        /// </summary>
+        /// <param name="devEUI">The device id.</param>
+        /// <param name="nwkSrvRx2Dr">The network server rx2 datarate.</param>
+        /// <param name="rx2DrFromTwins">RX2 datarate value from twins.</param>
+        /// <returns>The RX2 data rate.</returns>
+        public ushort GetDownstreamRX2DataRate(string devEUI, ushort? nwkSrvRx2Dr, ushort? rx2DrFromTwins, ILogger logger, DeviceJoinInfo deviceJoinInfo = null)
```
Comment (MaggieSalak) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoraTools/Regions/Region.cs:277]: This got deleted accidentally in PR https://github.com/Azure/iotedge-lorawan-starterkit/commit/e2b1b7bac49c7cf9a474c50ddef6158c150d05d7#diff-2890d1a0249a453790265b3dfd66d4661eb35fcc79183e05ae8ec16bd78bf0e6

File: Tests/Unit/LoRaTools/RegionTest.cs [hunk 2]
```diff
@@ -84,10 +90,25 @@ public void TestDownstreamRX2Frequency(Region region, double? nwksrvrx2freq, dou
             Assert.Equal(expectedFreq, freq);
         }
 
+        [Theory]
+        [MemberData(nameof(RegionEU868TestData.TestDownstreamRX2DataRateData), MemberType = typeof(RegionEU868TestData))]
+        [MemberData(nameof(RegionUS915TestData.TestDownstreamRX2DataRateData), MemberType = typeof(RegionUS915TestData))]
+        [MemberData(nameof(RegionCN470RP1TestData.TestDownstreamRX2DataRateData), MemberType = typeof(RegionCN470RP1TestData))]
+        [MemberData(nameof(RegionCN470RP2TestData.TestDownstreamRX2DataRateData), MemberType = typeof(RegionCN470RP2TestData))]
+        [MemberData(nameof(RegionAS923TestData.TestDownstreamRX2DataRateData), MemberType = typeof(RegionAS923TestData))]
+        public void TestDownstreamRX2DataRate(Region region, ushort? nwksrvrx2dr, ushort? rx2drfromtwins, ushort expectedDr, int? reportedJoinChannel = null, int? desiredJoinChannel = null)
```
Comment (MaggieSalak) [inline Tests/Unit/LoRaTools/RegionTest.cs:99]: Again, this was deleted by accident in PR https://github.com/Azure/iotedge-lorawan-starterkit/commit/e2b1b7bac49c7cf9a474c50ddef6158c150d05d7#diff-4762bf50d3ee4a29e4ae67282a6cac6512e69662af8215f0278229ea9864b794; this is just bringing it back.

File: Tests/Unit/LoRaTools/RegionTestBaseRxpk.cs [hunk 3]
```diff
@@ -125,5 +125,18 @@ protected void TestTryGetJoinChannelIndexRxpk(string dr, double freq, int expect
             Assert.False(Region.TryGetJoinChannelIndex(rxpk[0], out var channelIndex));
             Assert.Equal(expectedIndex, channelIndex);
         }
+
+        protected void TestIsDRValueWithinAcceptableValues(string dataRate, bool upstream, bool isValid)
```
Comment (MaggieSalak) [inline Tests/Unit/LoRaTools/RegionTestBaseRxpk.cs:129]: Adding coverage for untested methods `IsCurrentUpstreamDRValueWithinAcceptableValue` and ` IsCurrentDownstreamDRValueWithinAcceptableValue`.

File: Arduino/CN470RP1/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino [hunk 4]
```diff
@@ -0,0 +1,111 @@
+
+#include <LoRaWan.h>
+// THOSE ARDUINO SAMPLE REQUIRES CHANGES IN THE ARDUINO CODE TO RUN DESCRIBED IN THE MAIN REPO
```
Comment (Mandur) [inline Arduino/CN470RP1/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:3]: @p-schuler @MaggieSalak please review this file as I added it

File: Arduino/CN470RP1/TransmissionTestOTAALoRa/TransmissionTestOTAALoRa.ino [hunk 5]
```diff
@@ -0,0 +1,109 @@
+
+#include <LoRaWan.h>
+// THOSE ARDUINO SAMPLE REQUIRES CHANGES IN THE ARDUINO CODE TO RUN DESCRIBED IN THE MAIN REPO
```
Comment (Mandur) [inline Arduino/CN470RP1/TransmissionTestOTAALoRa/TransmissionTestOTAALoRa.ino:3]: @p-schuler @MaggieSalak please review this file as I added it

**Review by Mandur — APPROVED** [review]

**Comment by Mandur** [comment]
Very good, tested and works!

File: Arduino/CN470RP1/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    "NwkSKey": "3B7E151628AED2A6ABF7158809CF4F3C",
+    "DevAddr": "0228B1B1",
+    "GatewayID" :"",
+    "SensorDecoder" :"DecoderValueSensor"
+    },
+*/
+
+//set initial datarate and physical information for the device
+_data_rate_t dr = DR6;
+_physical_type_t physicalType = EU868 ;
```
Comment (MaggieSalak) [inline Arduino/CN470RP1/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:30]: @Mandur should this be EU868?

File: Arduino/CN470RP1/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+  lora.setId(devAddr, deviceId, NULL);
+  lora.setKey(nwkSKey, appSKey, NULL);
+
+  lora.setDeciveMode(LWABP);
+  lora.setDataRate(dr, physicalType);
+
+  // Disable Channels not used from the active channel list
+  for (int i = 4 ; i< 39; i++){
+    lora.setChannelOFF(i);
```
Comment (MaggieSalak) [inline Arduino/CN470RP1/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:54]: @Mandur `setChannelOFF()` doesn't seem to compile correctly, maybe we missed an import?

C:\Users\masalak\code\iotedge-lorawan-starterkit\Arduino\CN470RP1\TransmissionTestABPLoRa\TransmissionTestABPLoRa.ino: In function 'void setup()':
TransmissionTestABPLoRa:53:10: error: 'class LoRaWanClass' has no member named 'setChannelOFF'; did you mean 'setChannel'?
     lora.setChannelOFF(i);
          ^~~~~~~~~~~~~
          setChannel
TransmissionTestABPLoRa:56:10: error: 'class LoRaWanClass' has no member named 'setChannelOFF'; did you mean 'setChannel'?
     lora.setChannelOFF(i);
          ^~~~~~~~~~~~~
          setChannel
exit status 1
'class LoRaWanClass' has no member named 'setChannelOFF'; did you mean 'setChannel'?

File: Arduino/CN470RP1/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino (hunk 4 above)
Comment (MaggieSalak) [reply Arduino/CN470RP1/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:3]: Just reviewed. Looks good, I commited small blank line and identation fixes; I also added a comment about `setChannelOFF()` which doesn't compile for me in the Arduino IDE.

Comment (Mandur) [reply Arduino/CN470RP1/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:3]: Yes, you need a change in the arduino library, tracked in #1013

File: Arduino/CN470RP1/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino (hunk 6 above)
Comment (Mandur) [reply Arduino/CN470RP1/TransmissionTestABPLoRa/TransmissionTestABPLoRa.ino:30]: Oww must have been saving mistakes, Will fix thank you!

**Pull request merged**